src/steadystate_adjoint.jl

```diff
@@ -1,81 +1,170 @@
+    )
 
-  sense = SteadyStateAdjointSensitivityFunction(g,sensealg,discrete,sol,dg,f.colorvec)
-  @unpack diffcache, y, sol, λ, vjp = sense
-  if DiffEqBase.has_jac(f)
-     f.jac(diffcache.J,y,p,nothing)
-  else
-     if DiffEqBase.isinplace(sol.prob)
-       jacobian!(diffcache.J, diffcache.uf, y, diffcache.f_cache, sensealg, diffcache.jac_config)
-     else
-        temp = jacobian(diffcache.uf, y, sensealg)
-        @. diffcache.J = temp
-     end
-  end
+    sense = SteadyStateAdjointSensitivityFunction(
+        g,
+        sensealg,
+        discrete,
+        sol,
+        dg,
+        f.colorvec,
+        needs_jac,
+    )
+    @unpack diffcache, y, sol, λ, vjp, linsolve = sense
 
-  _save_idxs = save_idxs === nothing ? Colon() : save_idxs
-  if dg !== nothing
-    if g!== nothing
-      dg(vec(diffcache.dg_val),y,p,nothing,nothing)
-    else
-      if typeof(_save_idxs) <: Number
-        diffcache.dg_val[_save_idxs] = dg[_save_idxs]
-      elseif typeof(dg) <: Number
-        @. diffcache.dg_val[_save_idxs] = dg
-      else
-        @. diffcache.dg_val[_save_idxs] = dg[_save_idxs]
-      end
+    if needs_jac
+        if DiffEqBase.has_jac(f)
+            f.jac(diffcache.J, y, p, nothing)
+        else
+            if DiffEqBase.isinplace(sol.prob)
+                jacobian!(
+                    diffcache.J,
+                    diffcache.uf,
+                    y,
+                    diffcache.f_cache,
+                    sensealg,
+                    diffcache.jac_config,
+                )
+            else
+                temp = jacobian(diffcache.uf, y, sensealg)
+                @. diffcache.J = temp
+            end
+        end
     end
-  else
-    if g !== nothing
-      gradient!(vec(diffcache.dg_val),diffcache.g,y,sensealg,diffcache.g_grad_config)
+
+    _save_idxs = save_idxs === nothing ? Colon() : save_idxs
+    if dg !== nothing
+        if g !== nothing
+            dg(vec(diffcache.dg_val), y, p, nothing, nothing)
+        else
+            if typeof(_save_idxs) <: Number
+                diffcache.dg_val[_save_idxs] = dg[_save_idxs]
+            elseif typeof(dg) <: Number
+                @. diffcache.dg_val[_save_idxs] = dg
+            else
+                @. diffcache.dg_val[_save_idxs] = dg[_save_idxs]
+            end
+        end
+    else
+        if g !== nothing
+            gradient!(
+                vec(diffcache.dg_val),
+                diffcache.g,
+                y,
+                sensealg,
+                diffcache.g_grad_config,
+            )
+        end
     end
-  end
 
-  copyto!(vec(λ), diffcache.J'\vec(diffcache.dg_val')) # use linsolve here
-  vecjacobian!(vec(diffcache.dg_val), y, λ, p, nothing, sense, dgrad=vjp, dy=nothing)
+    if !needs_jac
+        # NOTE: Zygote doesn't support inplace
+        linear_problem = LinearProblem(VecJacOperator(f, y, p; autodiff = !DiffEqBase.isinplace(sol.prob)),
```

Comment:
We should make this use an internal VecJacOperator that has all of the vjp choices matching the ODE cases. But we can keep that for a future PR.